File: code_puppy/config/debug.py
"""Feature toggles, debug flags, and environment configuration.

Mirrors ``CodePuppyControl.Config.Debug`` in the Elixir runtime.

Centralizes all boolean feature flags and debug-related settings from
``puppy.cfg``. Each getter has a documented default that matches the
Python/Elixir behavior.

Config keys in puppy.cfg:

- ``yolo_mode`` — auto-approve all actions (default true)
- ``allow_recursion`` — allow recursive agent calls (default true)
- ``enable_dbos`` — enable DBOS workflow engine (default true)
- ``enable_pack_agents`` — enable pack agents (default false)
- ``enable_universal_constructor`` — enable dynamic tool creation (default true)
- ``enable_streaming`` — enable SSE streaming (default true)
- ``enable_agent_memory`` — enable cross-session agent memory (default false)
- ``http2`` — enable HTTP/2 for httpx clients (default false)
- ``subagent_verbose`` — verbose output for sub-agents (default false)
- ``disable_mcp`` — skip MCP server loading (default false)
- ``safety_permission_level`` — risk threshold (default ``"medium"``)
- ``enable_user_plugins`` — enable user plugin loading (default true)
- ``allowed_user_plugins`` — comma-separated plugin allowlist
"""

from __future__ import annotations

import os
from pathlib import Path

from code_puppy.config.loader import (
    _is_truthy,
    _make_bool_getter,
    _make_int_getter,
    _registered_cache,
    get_value,
    set_config_value,
)

__all__ = [
    "get_yolo_mode",
    "get_allow_recursion",
    "get_use_dbos",
    "set_enable_dbos",
    "get_pack_agents_enabled",
    "PACK_AGENT_NAMES",
    "UC_AGENT_NAMES",
    "get_universal_constructor_enabled",
    "set_universal_constructor_enabled",
    "get_enable_streaming",
    "get_enable_agent_memory",
    "get_adaptive_rendering_enabled",
    "get_post_edit_validation_enabled",
    "get_subagent_verbose",
    "get_http2",
    "set_http2",
    "get_mcp_disabled",
    "get_safety_permission_level",
    "get_enable_user_plugins",
    "get_allowed_user_plugins",
    "load_api_keys_to_environment",
    "get_api_key",
    "set_api_key",
    # ADR-003 / feature toggles
    "get_elixir_message_shadow_mode_enabled",
    "get_enable_gitignore_filtering",
]


# ---------------------------------------------------------------------------
# YOLO mode
# ---------------------------------------------------------------------------


@_registered_cache
def get_yolo_mode() -> bool:
    """Return True if YOLO mode is enabled (auto-approve actions, default true)."""
    return _is_truthy(get_value("yolo_mode"), default=True)


# ---------------------------------------------------------------------------
# Recursion
# ---------------------------------------------------------------------------


get_allow_recursion = _make_bool_getter(
    "allow_recursion",
    default=True,
    doc="Return True if recursive agent calls are allowed (default True).",
)


# ---------------------------------------------------------------------------
# DBOS
# ---------------------------------------------------------------------------


@_registered_cache
def get_use_dbos() -> bool:
    """Return True if DBOS should be used.

    Returns True only when BOTH conditions are met:
    1. ``enable_dbos`` is not explicitly set to false in puppy.cfg (default: true)
    2. The dbos package is actually installed
    """
    if not _is_truthy(get_value("enable_dbos"), default=True):
        return False
    try:
        import dbos as _dbos  # noqa: F401

        return True
    except ImportError:
        return False


def set_enable_dbos(enabled: bool) -> None:
    """Enable or disable DBOS via config."""
    set_config_value("enable_dbos", "true" if enabled else "false")


# ---------------------------------------------------------------------------
# Pack agents
# ---------------------------------------------------------------------------

PACK_AGENT_NAMES = frozenset(
    ["pack-leader", "shepherd", "terrier", "watchdog", "retriever"]
)

UC_AGENT_NAMES = frozenset(["helios"])

get_pack_agents_enabled = _make_bool_getter(
    "enable_pack_agents",
    default=False,
    doc="""Return True if pack agents are enabled (default False).""",
)


# ---------------------------------------------------------------------------
# Universal Constructor
# ---------------------------------------------------------------------------


get_universal_constructor_enabled = _make_bool_getter(
    "enable_universal_constructor",
    default=True,
    doc="""Return True if the Universal Constructor is enabled (default True).""",
)


def set_universal_constructor_enabled(enabled: bool) -> None:
    """Enable or disable the Universal Constructor."""
    set_config_value("enable_universal_constructor", "true" if enabled else "false")


# ---------------------------------------------------------------------------
# Streaming
# ---------------------------------------------------------------------------


get_enable_streaming = _make_bool_getter(
    "enable_streaming",
    default=True,
    doc="Return True if SSE streaming is enabled (default True).",
)


# ---------------------------------------------------------------------------
# Agent memory
# ---------------------------------------------------------------------------


@_registered_cache
def get_enable_agent_memory() -> bool:
    """Return True if agent memory is enabled (default False).

    DEPRECATED(audit-2026): Use memory_enabled instead.
    """
    return _is_truthy(get_value("enable_agent_memory"), default=False)


get_memory_debounce_seconds = _make_int_getter(
    "memory_debounce_seconds",
    default=30,
    min_val=1,
    max_val=300,
    doc="Return memory write debounce in seconds (default 30).",
)

get_memory_max_facts = _make_int_getter(
    "memory_max_facts",
    default=50,
    min_val=1,
    max_val=1000,
    doc="Return max facts per agent (default 50).",
)

get_memory_token_budget = _make_int_getter(
    "memory_token_budget",
    default=500,
    min_val=100,
    max_val=2000,
    doc="Return token budget for memory injection (default 500).",
)


@_registered_cache
def get_memory_extraction_model() -> str | None:
    """Return the optional model override for memory extraction."""
    return get_value("memory_extraction_model")


# ---------------------------------------------------------------------------
# Adaptive rendering
# ---------------------------------------------------------------------------


@_registered_cache
def get_adaptive_rendering_enabled() -> bool:
    """Return True if adaptive payload rendering is enabled (default True)."""
    from code_puppy.config_package.env_helpers import env_bool

    env_val = env_bool("PUPPY_ADAPTIVE_RENDERING", default=True)
    if not env_val:
        return False
    return _is_truthy(get_value("adaptive_rendering"), default=True)


# ---------------------------------------------------------------------------
# Post-edit validation
# ---------------------------------------------------------------------------


@_registered_cache
def get_post_edit_validation_enabled() -> bool:
    """Return True if post-edit syntax validation is enabled (default True)."""
    from code_puppy.config_package.env_helpers import env_bool

    env_val = env_bool("PUPPY_POST_EDIT_VALIDATION", default=True)
    if not env_val:
        return False
    return _is_truthy(get_value("enable_post_edit_validation"), default=True)


# ---------------------------------------------------------------------------
# Sub-agent verbose
# ---------------------------------------------------------------------------


get_subagent_verbose = _make_bool_getter(
    "subagent_verbose",
    default=False,
    doc="Return True if sub-agent verbose output is enabled (default False).",
)


# ---------------------------------------------------------------------------
# HTTP/2
# ---------------------------------------------------------------------------


get_http2 = _make_bool_getter(
    "http2",
    default=False,
    doc="Return True if HTTP/2 is enabled (default False).",
)


def set_http2(enabled: bool) -> None:
    """Set the http2 configuration value."""
    set_config_value("http2", "true" if enabled else "false")


# ---------------------------------------------------------------------------
# MCP disabled
# ---------------------------------------------------------------------------


get_mcp_disabled = _make_bool_getter(
    "disable_mcp",
    default=False,
    doc="Return True if MCP is disabled (default False).",
)


# ---------------------------------------------------------------------------
# Safety permission level
# ---------------------------------------------------------------------------


_VALID_SAFETY_LEVELS = frozenset({"none", "low", "medium", "high", "critical"})


@_registered_cache
def get_safety_permission_level() -> str:
    """Return the safety permission level (default ``"medium"``)."""
    val = get_value("safety_permission_level")
    if val:
        normalized = str(val).strip().lower()
        if normalized in _VALID_SAFETY_LEVELS:
            return normalized
    return "medium"


# ---------------------------------------------------------------------------
# User plugin security
# ---------------------------------------------------------------------------


get_enable_user_plugins = _make_bool_getter(
    "enable_user_plugins",
    default=True,
    doc="Return True if user plugin loading is enabled (default True).",
)


@_registered_cache
def get_allowed_user_plugins() -> list[str]:
    """Return the list of allowed user plugin names from config."""
    val = get_value("allowed_user_plugins")
    if not val:
        return []
    return [name.strip() for name in val.split(",") if name.strip()]


# ---------------------------------------------------------------------------
# API key management
# ---------------------------------------------------------------------------


def get_api_key(key_name: str) -> str:
    """Get an API key from puppy.cfg."""
    return get_value(key_name) or ""


def set_api_key(key_name: str, value: str) -> None:
    """Set an API key in puppy.cfg."""
    set_config_value(key_name, value)


# ---------------------------------------------------------------------------
# Elixir shadow mode
# ---------------------------------------------------------------------------


get_elixir_message_shadow_mode_enabled = _make_bool_getter(
    "enable_elixir_message_shadow_mode",
    default=False,
    doc="""Return True if Elixir message shadow mode is enabled (default False).

    When enabled, both Python and Elixir execute message operations
    and results are compared for divergence detection.
    """,
)


# ---------------------------------------------------------------------------
# Gitignore filtering
# ---------------------------------------------------------------------------


get_enable_gitignore_filtering = _make_bool_getter(
    "enable_gitignore_filtering",
    default=False,
    doc="""Return True if gitignore-based file filtering is enabled (default False).

    When enabled, file operations respect .gitignore patterns.
    Higher-risk flag: requires explicit opt-in.
    """,
)


# Code Puppy environment variable allowlist
_CODEPUPPY_ENV_ALLOWLIST: set[str] = {
    "ANTHROPIC_API_KEY",
    "OPENAI_API_KEY",
    "GEMINI_API_KEY",
    "GOOGLE_API_KEY",
    "CEREBRAS_API_KEY",
    "SYN_API_KEY",
    "AZURE_OPENAI_API_KEY",
    "AZURE_OPENAI_ENDPOINT",
    "OPENROUTER_API_KEY",
    "ZAI_API_KEY",
    "GITHUB_TOKEN",
    "FIREWORKS_API_KEY",
    "GROQ_API_KEY",
    "MISTRAL_API_KEY",
    "MOONSHOT_API_KEY",
    "PUP_DEBUG",
    "PUP_MODEL",
    "PUP_AGENT",
    "PUPPY_DEFAULT_AGENT",
    "PUPPY_DEFAULT_MODEL",
    "PUPPY_TEMPERATURE",
    "PUPPY_MESSAGE_LIMIT",
    "PUPPY_PROTECTED_TOKEN_COUNT",
    "PUP_DISABLE_CALLBACK_PLUGIN_LOADING",
    "CODE_PUPPY_SKIP_TUTORIAL",
    "CODE_PUPPY_NO_TUI",
    "CODE_PUPPY_NO_COLOR",
    "CODE_PUPPY_FORCE_COLOR",
    "CODE_PUPPY_CODE_THEME",
    "CODE_PUPPY_ALLOWED_ORIGINS",
    "CODE_PUPPY_DISABLE_RETRY_TRANSPORT",
    "CODE_PUPPY_BRIDGE",
    "PUPPY_ADAPTIVE_RENDERING",
    "PUPPY_POST_EDIT_VALIDATION",
    "PUPPY_WS_HISTORY_TTL_SECONDS",
    "PUPPY_WS_HISTORY_MAXLEN",
    "CODE_PUPPY_BRIDGE_HOST",
    "NO_VERSION_UPDATE",
}


def _load_env_allowlisted(env_file: Path, allowlist: set[str]) -> None:
    """Load only allowlisted keys from .env file into os.environ."""
    if not env_file.exists():
        return
    try:
        from dotenv import dotenv_values
    except ImportError:
        return
    values = dotenv_values(env_file)
    for key in allowlist:
        if key in values and key not in os.environ:
            os.environ[key] = values[key]


def load_api_keys_to_environment() -> None:
    """Load all API keys from .env and puppy.cfg into environment variables.

    Priority: .env > puppy.cfg > existing env vars.
    Security: Only allowlisted Code Puppy env vars are loaded from .env.
    """
    api_key_names = [
        "OPENAI_API_KEY",
        "GEMINI_API_KEY",
        "ANTHROPIC_API_KEY",
        "CEREBRAS_API_KEY",
        "SYN_API_KEY",
        "AZURE_OPENAI_API_KEY",
        "AZURE_OPENAI_ENDPOINT",
        "OPENROUTER_API_KEY",
        "ZAI_API_KEY",
    ]

    env_file = Path.cwd() / ".env"
    _load_env_allowlisted(env_file, _CODEPUPPY_ENV_ALLOWLIST)

    for key_name in api_key_names:
        if key_name not in os.environ or not os.environ[key_name]:
            value = get_api_key(key_name)
            if value:
                os.environ[key_name] = value

File: elixir/code_puppy_control/lib/code_puppy_control/config/debug.ex
defmodule CodePuppyControl.Config.Debug do
  @moduledoc """
  Feature toggles, debug flags, and environment configuration.

  Centralizes all boolean feature flags and debug-related settings from
  `puppy.cfg`. Each getter has a documented default that matches the
  Python `config.py` behaviour.

  ## Config keys in `puppy.cfg`

  - `yolo_mode` — auto-approve all actions (default `true`)
  - `allow_recursion` — allow recursive agent calls (default `true`)
  - `enable_dbos` — enable DBOS workflow engine (default `true`)
  - `enable_pack_agents` — enable pack agents (default `false`)
  - `enable_universal_constructor` — enable dynamic tool creation (default `true`)
  - `enable_streaming` — enable SSE streaming (default `true`)
  - `enable_gitignore_filtering` — filter gitignored files in list_files (default `false`)
  - `enable_agent_memory` — enable cross-session agent memory (default `false`)
  - `http2` — enable HTTP/2 for httpx clients (default `false`)
  - `subagent_verbose` — verbose output for sub-agents (default `false`)
  - `disable_mcp` — skip MCP server loading (default `false`)
  - `grep_output_verbose` — full grep output (default `false`)
  - `safety_permission_level` — risk threshold (default `"medium"`)
  - `debug` — debug mode flag
  """

  alias CodePuppyControl.Config.Loader

  # ── Core toggles ────────────────────────────────────────────────────────

  @doc "Return `true` if yolo mode is on (default `true`)."
  @spec yolo_mode?() :: boolean()
  def yolo_mode?, do: truthy?("yolo_mode", true)

  @doc "Return `true` if recursion is allowed (default `true`)."
  @spec allow_recursion?() :: boolean()
  def allow_recursion?, do: truthy?("allow_recursion", true)

  @doc "Return `true` if DBOS is enabled (default `true`)."
  @spec dbos_enabled?() :: boolean()
  def dbos_enabled?, do: truthy?("enable_dbos", true)

  @doc "Return `true` if pack agents are enabled (default `false`)."
  @spec pack_agents_enabled?() :: boolean()
  def pack_agents_enabled?, do: truthy?("enable_pack_agents", false)

  @doc "Return `true` if the Universal Constructor is enabled (default `true`)."
  @spec universal_constructor_enabled?() :: boolean()
  def universal_constructor_enabled?, do: truthy?("enable_universal_constructor", true)

  @doc "Return `true` if streaming is enabled (default `true`)."
  @spec streaming_enabled?() :: boolean()
  def streaming_enabled?, do: truthy?("enable_streaming", true)

  @doc "Return `true` if gitignore filtering is enabled (default `false`)."
  @spec gitignore_filtering_enabled?() :: boolean()
  def gitignore_filtering_enabled?, do: truthy?("enable_gitignore_filtering", false)

  @doc """
  Return `true` if Elixir message shadow mode is enabled (default `false`).

  When enabled, both Python and Elixir execute the same message operation
  and results are compared for divergence detection. Opt-in only.
  Config key: `enable_elixir_message_shadow_mode`
  """
  @spec elixir_message_shadow_mode_enabled?() :: boolean()
  def elixir_message_shadow_mode_enabled?, do: truthy?("enable_elixir_message_shadow_mode", false)

  @doc "Return `true` if agent memory is enabled (default `false`)."
  @spec agent_memory_enabled?() :: boolean()
  def agent_memory_enabled?, do: truthy?("enable_agent_memory", false)

  @doc "Return `true` if skills integration is enabled (default `true`)."
  @spec skills_enabled?() :: boolean()
  def skills_enabled?, do: truthy?("skills_enabled", true)

  @doc "Return `true` if HTTP/2 is enabled (default `false`)."
  @spec http2_enabled?() :: boolean()
  def http2_enabled?, do: truthy?("http2", false)

  @doc "Return `true` if sub-agent verbose output is enabled (default `false`)."
  @spec subagent_verbose?() :: boolean()
  def subagent_verbose?, do: truthy?("subagent_verbose", false)

  @doc "Return `true` if MCP is disabled (default `false`)."
  @spec mcp_disabled?() :: boolean()
  def mcp_disabled?, do: truthy?("disable_mcp", false)

  @doc """
  Adaptive rendering enabled.
  Env override: PUP_ADAPTIVE_RENDERING
  """
  def adaptive_rendering_enabled? do
    case System.get_env("PUP_ADAPTIVE_RENDERING") do
      nil -> truthy?("adaptive_rendering", true)
      val -> val not in ["0", "false", "no"]
    end
  end

  @doc """
  Post-edit validation enabled.
  Env override: PUP_POST_EDIT_VALIDATION
  """
  def post_edit_validation_enabled? do
    case System.get_env("PUP_POST_EDIT_VALIDATION") do
      nil -> truthy?("post_edit_validation", true)
      val -> val not in ["0", "false", "no"]
    end
  end

  # ── Setters ─────────────────────────────────────────────────────────────

  @doc "Enable or disable the Universal Constructor."
  @spec set_universal_constructor_enabled(boolean()) :: :ok
  def set_universal_constructor_enabled(enabled) do
    CodePuppyControl.Config.Writer.set_value("enable_universal_constructor", bool_str(enabled))
  end

  @doc "Enable or disable DBOS."
  @spec set_dbos_enabled(boolean()) :: :ok
  def set_dbos_enabled(enabled) do
    CodePuppyControl.Config.Writer.set_value("enable_dbos", bool_str(enabled))
  end

  @doc "Enable or disable HTTP/2."
  @spec set_http2_enabled(boolean()) :: :ok
  def set_http2_enabled(enabled) do
    CodePuppyControl.Config.Writer.set_value("http2", bool_str(enabled))
  end

  # ── Safety ──────────────────────────────────────────────────────────────

  @valid_safety_levels MapSet.new(["none", "low", "medium", "high", "critical"])

  @doc """
  Return the safety permission level (default `"medium"`).
  Valid values: `none`, `low`, `medium`, `high`, `critical`.
  """
  @spec safety_permission_level() :: String.t()
  def safety_permission_level do
    case Loader.get_value("safety_permission_level") do
      nil ->
        "medium"

      val ->
        normalized = String.downcase(String.trim(val))
        if normalized in @valid_safety_levels, do: normalized, else: "medium"
    end
  end

  # ── Debug mode ──────────────────────────────────────────────────────────

  @doc """
  Return `true` if debug mode is active.

  Checks `PUP_DEBUG` env var first, then `debug` key in `puppy.cfg`.
  """
  @spec debug?() :: boolean()
  def debug? do
    case System.get_env("PUP_DEBUG") do
      nil -> truthy?("debug", false)
      val -> val != "" and val != "0"
    end
  end

  # ── Token / puppy token ─────────────────────────────────────────────────

  @doc "Return the puppy_token from config, or `nil`."
  @spec puppy_token() :: String.t() | nil
  def puppy_token do
    Loader.get_value("puppy_token")
  end

  @doc "Set the puppy_token in config."
  @spec set_puppy_token(String.t()) :: :ok
  def set_puppy_token(token) when is_binary(token) do
    CodePuppyControl.Config.Writer.set_value("puppy_token", token)
  end

  # ── Memory config ───────────────────────────────────────────────────────

  @doc "Return memory write debounce in seconds (default `30`, range `1–300`)."
  @spec memory_debounce_seconds() :: pos_integer()
  def memory_debounce_seconds do
    case Loader.get_value("memory_debounce_seconds") do
      nil ->
        30

      val ->
        case Integer.parse(val) do
          {n, _} -> n |> max(1) |> min(300)
          :error -> 30
        end
    end
  end

  @doc "Return max facts per agent (default `50`, range `1–1000`)."
  @spec memory_max_facts() :: pos_integer()
  def memory_max_facts do
    case Loader.get_value("memory_max_facts") do
      nil ->
        50

      val ->
        case Integer.parse(val) do
          {n, _} -> n |> max(1) |> min(1000)
          :error -> 50
        end
    end
  end

  @doc "Return token budget for memory injection (default `500`, range `100–2000`)."
  @spec memory_token_budget() :: pos_integer()
  def memory_token_budget do
    case Loader.get_value("memory_token_budget") do
      nil ->
        500

      val ->
        case Integer.parse(val) do
          {n, _} -> n |> max(100) |> min(2000)
          :error -> 500
        end
    end
  end

  # ── API keys ────────────────────────────────────────────────────────────

  @api_key_names [
    "OPENAI_API_KEY",
    "GEMINI_API_KEY",
    "ANTHROPIC_API_KEY",
    "CEREBRAS_API_KEY",
    "SYN_API_KEY",
    "AZURE_OPENAI_API_KEY",
    "AZURE_OPENAI_ENDPOINT",
    "OPENROUTER_API_KEY",
    "ZAI_API_KEY",
    "WAFER_API_KEY"
  ]

  @doc """
  Load API keys from `puppy.cfg` into environment variables.

  Only sets env vars that are not already present, preserving `.env` and
  shell-level overrides.
  """
  @spec load_api_keys_to_environment() :: :ok
  def load_api_keys_to_environment do
    Enum.each(@api_key_names, fn key ->
      if is_nil(System.get_env(key)) or System.get_env(key) == "" do
        # Config keys are lowercase (e.g., "wafer_api_key")
        config_key = String.downcase(key)

        case Loader.get_value(config_key) do
          nil -> :ok
          "" -> :ok
          value -> System.put_env(key, value)
        end
      end
    end)

    :ok
  end

  # ── Private ─────────────────────────────────────────────────────────────

  @truthy_values MapSet.new(["1", "true", "yes", "on"])

  defp truthy?(key, default) do
    case Loader.get_value(key) do
      nil -> default
      val -> String.downcase(String.trim(val)) in @truthy_values
    end
  end

  defp bool_str(true), do: "true"
  defp bool_str(false), do: "false"
end
